actors/market/src/lib.rs

```diff
@@ -531,92 +531,182 @@ impl Actor {
+        params: BatchActivateDealsParams,
+    ) -> Result<BatchActivateDealsResult, ActorError> {
         rt.validate_immediate_caller_type(std::iter::once(&Type::Miner))?;
         let miner_addr = rt.message().caller();
         let curr_epoch = rt.curr_epoch();
 
-        let (deal_spaces, verified_infos) = rt.transaction(|st: &mut State, rt| {
-            let proposal_array = st.get_proposal_array(rt.store())?;
-            let proposals = get_proposals(&proposal_array, &params.deal_ids, st.next_id)?;
-
-            let deal_spaces = {
-                validate_and_return_deal_space(
-                    &proposals,
-                    &miner_addr,
-                    params.sector_expiry,
-                    curr_epoch,
-                    None,
-                )
-                .context("failed to validate deal proposals for activation")?
-            };
-
-            // Update deal states
-            let mut verified_infos = Vec::new();
+        let sector_results = rt.transaction(|st: &mut State, rt| {
             let mut deal_states: Vec<(DealID, DealState)> = vec![];
 
-            for (deal_id, proposal) in proposals {
-                // This construction could be replaced with a single "update deal state"
-                // state method, possibly batched over all deal ids at once.
-                let s = st.find_deal_state(rt.store(), deal_id)?;
-
-                if s.is_some() {
-                    return Err(actor_error!(
-                        illegal_argument,
-                        "deal {} already activated",
-                        deal_id
-                    ));
-                }
+            let sector_results: Vec<Option<ActivateDealsResult>> = params
+                .sectors
+                .iter()
+                .map(|p| {
+                    if p.deal_ids.is_empty() {
+                        return Some(ActivateDealsResult {
+                            nonverified_deal_space: BigInt::default(),
+                            verified_infos: Vec::default(),
+                        });
+                    }
 
-                let propc = rt_deal_cid(rt, &proposal)?;
+                    let proposal_array = match st.get_proposal_array(rt.store()) {
+                        Ok(proposal_array) => proposal_array,
+                        Err(e) => {
+                            log::warn!("failed to activate deals {:?}: {}", p.deal_ids, e);
+                            return None;
+                        }
+                    };
+                    let proposals = match get_proposals(&proposal_array, &p.deal_ids, st.next_id) {
+                        Ok(proposals) => proposals,
+                        Err(e) => {
+                            log::warn!("failed to activate deals {:?}: {}", p.deal_ids, e);
+                            return None;
+                        }
+                    };
+
+                    let deal_spaces = match validate_and_return_deal_space(
```

Comment:
There's potentially a major security issue here.
`validate_and_return_deal_spaces` needs access to all deals at once to make sure we're not double activating deals and over claiming QA power.  But as I understand this change this is being called once per sector.  A concrete attack on this is a batch of n deal activations all over the same deal ids giving the caller an amplification of n times their qa power.

It might be that verified registry now saves us here, will have to look more.

We should have a test of this case as part of this change.

Reply:
Ok it looks like the duplicates will be fed to `CLAIM_ALLOCATIONS` which doesn't have the same bug and will just fail the whole thing ✅ 
That said we should still avoid double activations on market actor or convince ourselves that activation is idempotent and then remove all duplication checks.

Reply:
I've made changes so that if a duplicate deal is detected between sectors, the subsequent activations are skipped and marked as failed. This preserves the behaviour checked in `replica_update_test::dealincluded_in_multiple_sectors_failure`

Reply:
> replica_update_test::dealincluded_in_multiple_sectors_failure

> It's get_proposals that used to check for duplicates

`validate_deal_can_activate` checks for duplicates that are introduced by PSD parameters which is the case I was originally raising.

> This preserves the behaviour checked in replica_update_test::dealincluded_in_multiple_sectors_failure

Looking at this test I don't think it exercises the case I'm describing.  The problem is not about a deal already existing in a sector but rather that a PSD message can activate the same deal twice.  I think either there is some other existing test or you'll need to write a new one.